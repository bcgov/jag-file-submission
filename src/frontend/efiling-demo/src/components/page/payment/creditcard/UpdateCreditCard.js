import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { Footer, Header, Button, Table } from "shared-components";
import { useLocation } from "react-router-dom";
import queryString from "query-string";

import { propTypes } from "../../../../types/propTypes";

export default function UpdateCreditCard({ page: { header } }) {
  const successCode = 1;
  const errorCode = 19;

  const [queryParams, setqueryParams] = useState([]);
  const [redirectUrl, setRedirectUrl] = useState(null);
  const location = useLocation();

  const redirect = (code) => {
    window.open(`${redirectUrl}?responseCode=${code}`, "_self");
  };

  useEffect(() => {
    const paramters = [];

    const queryStrings = queryString.parse(location.search);

    Object.keys(queryStrings).forEach((key) => {
      paramters.push({
        name: key,
        value: queryStrings[key],
        isValueBold: true,
      });
    });

    setRedirectUrl(queryStrings.trnReturnURL);
    setqueryParams(paramters);
  }, []);

  return (
    <main>
      <Header header={header} />
      <div className="page">
        <div className="content container">
          <div className="row justify-content-md-center">
            <div className="col-md-8 col-sm-12">
              <div className="card">
                <div className="card-body">
                  <h5 className="card-title">Query String values</h5>
                  <Table elements={queryParams} />
                </div>
              </div>
            </div>
          </div>
          <br />
          <div className="row justify-content-md-center">
            <div className="col-md-4 col-sm-12 mt-2">
              <Button
                onClick={() => {
                  redirect(successCode);
                }}
                label="Simulate Card Update Success"
                styling="btn btn-success mx-auto"
              />
            </div>
            <div className="col-md-4 col-sm-12 mt-2">
              <Button
                onClick={() => {
                  redirect(errorCode);
                }}
                label="Simulate Card Update Failure"
                styling="btn btn-danger mx-auto"
              />
            </div>
          </div>
        </div>
      </div>

      <Footer />
    </main>
  );
}

UpdateCreditCard.propTypes = {
  page: PropTypes.shape({
    header: propTypes.header,
  }).isRequired,
};
