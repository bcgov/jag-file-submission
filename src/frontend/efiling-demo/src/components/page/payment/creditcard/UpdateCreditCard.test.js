import React from "react";
import { createMemoryHistory } from "history";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router";
import UpdateCreditCard from "./UpdateCreditCard";

describe("Success", () => {
  const header = {
    name: "eFiling Demo Client",
    history: createMemoryHistory(),
  };

  const page = { header };

  window.open = jest.fn();

  test("Component matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter
        initialEntries={[
          "/updateCreditCard?serviceVersion=1&merchantId=merchantid&trnLanguage=eng&operationType=N&ref1=customerCode&trnReturnURL=http://localhost:3000/efilinghub&trnOrderNumber=C1234&hashValue=97F9D18E87F902807890CBF8B11D3244&hashExpiry=202012302048",
        ]}
      >
        <Routes>
          <Route
            path="/updateCreditCard"
            element={<UpdateCreditCard page={page} />}
          />
        </Routes>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Click on success should redirect to efilinghub with error code 1", () => {
    global.open = jest.fn();

    const { getByText } = render(
      <MemoryRouter
        initialEntries={[
          "/updateCreditCard?serviceVersion=1&merchantId=merchantid&trnLanguage=eng&operationType=N&ref1=customerCode&trnReturnURL=http://localhost:3000/efilinghub&trnOrderNumber=C1234&hashValue=97F9D18E87F902807890CBF8B11D3244&hashExpiry=202012302048",
        ]}
      >
        <Routes>
          <Route
            path="/updateCreditCard"
            element={<UpdateCreditCard page={page} />}
          />
        </Routes>
      </MemoryRouter>
    );

    waitFor(() => getByText("Simulate Card Update Success"));
    fireEvent.click(getByText("Simulate Card Update Success"));

    expect(global.open).toBeCalledWith(
      "http://localhost:3000/efilinghub?responseCode=1",
      "_self"
    );
  });

  test("Click on failure should redirect to efilinghub with error code 19", () => {
    global.open = jest.fn();

    const { getByText } = render(
      <MemoryRouter
        initialEntries={[
          "/updateCreditCard?serviceVersion=1&merchantId=merchantid&trnLanguage=eng&operationType=N&ref1=customerCode&trnReturnURL=http://localhost:3000/efilinghub&trnOrderNumber=C1234&hashValue=97F9D18E87F902807890CBF8B11D3244&hashExpiry=202012302048",
        ]}
      >
        <Routes>
          <Route
            path="/updateCreditCard"
            element={<UpdateCreditCard page={page} />}
          />
        </Routes>
      </MemoryRouter>
    );
    waitFor(() => getByText("Simulate Card Update Failure"));
    fireEvent.click(getByText("Simulate Card Update Failure"));

    expect(global.open).toBeCalledWith("null?responseCode=19", "_self");
  });
});
