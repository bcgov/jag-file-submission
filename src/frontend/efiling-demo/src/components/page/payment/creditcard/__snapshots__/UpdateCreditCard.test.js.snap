// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Success Component matches the snapshot 1`] = `
<DocumentFragment>
  <main>
    <header
      class="bcgov-header"
    >
      <nav
        aria-label="Header"
        class="container-fluid navbar navbar-expand-lg navbar-dark"
      >
        <div
          aria-labelledby="title"
          class="navbar-brand bcgov-pointer"
        />
        <img
          alt="B.C. Government Logo"
          class="img-fluid d-none d-md-block bcgov-pointer"
          height="44"
          src="undefined/images/bcid-logo-rev-en.svg"
          width="181"
        />
        <img
          alt="B.C. Government Logo"
          class="img-fluid d-md-none bcgov-pointer"
          height="44"
          src="undefined/images/bcid-symbol-rev.svg"
          width="64"
        />
        <div
          aria-labelledby="title"
          class="bcgov-pointer navbar-brand nav-item nav-link"
        />
        <div
          class="bcgov-flex-spacing"
        >
          <div
            class="navbar-brand bcgov-vertical-center"
            id="title"
          >
            eFiling Demo Client
          </div>
          <div />
        </div>
      </nav>
    </header>
    <div
      class="page"
    >
      <div
        class="content container"
      >
        <div
          class="row justify-content-md-center"
        >
          <div
            class="col-md-8 col-sm-12"
          >
            <div
              class="card"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Query String values
                </h5>
                <div
                  class="bcgov-table "
                >
                  <b />
                  <div
                    class="bcgov-table-body"
                  >
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        hashExpiry
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          202012302048
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        hashValue
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          97F9D18E87F902807890CBF8B11D3244
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        merchantId
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          merchantid
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        operationType
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          N
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        ref1
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          customerCode
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        serviceVersion
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          1
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        trnLanguage
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          eng
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        trnOrderNumber
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          C1234
                        </b>
                      </div>
                    </div>
                    <div
                      class="bcgov-row "
                    >
                      <div
                        class="bcgov-fill-width"
                      >
                        trnReturnURL
                      </div>
                      <div
                        class="bcgov-table-value "
                      >
                        <b>
                          http://localhost:3000/efilinghub
                        </b>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <br />
        <div
          class="row justify-content-md-center"
        >
          <div
            class="col-md-4 col-sm-12 mt-2"
          >
            <button
              class="bcgov-button btn btn-success mx-auto"
              data-test-id=""
              type="button"
            >
              Simulate Card Update Success
            </button>
          </div>
          <div
            class="col-md-4 col-sm-12 mt-2"
          >
            <button
              class="bcgov-button btn btn-danger mx-auto"
              data-test-id=""
              type="button"
            >
              Simulate Card Update Failure
            </button>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="bcgov-footer"
    >
      <nav
        aria-label="Footer"
        class="navbar navbar-expand-sm navbar-dark justify-content-end"
      >
        <button
          class="navbar-toggler"
          data-target="#footerBar"
          data-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse flex-grow-0"
          id="footerBar"
        >
          <ul
            class="navbar-nav text-right"
          >
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca"
                rel="noopener noreferrer"
                target="_blank"
              >
                BC Government
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/disclaimer"
                rel="noopener noreferrer"
                target="_blank"
              >
                Disclaimer
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/privacy"
                rel="noopener noreferrer"
                target="_blank"
              >
                Privacy
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/accessibility"
                rel="noopener noreferrer"
                target="_blank"
              >
                Accessibility
              </a>
            </li>
            <li
              class="nav-item m-1"
            >
              <a
                class="nav-link"
                href="https://www2.gov.bc.ca/gov/content/home/copyright"
                rel="noopener noreferrer"
                target="_blank"
              >
                Copyright
              </a>
            </li>
          </ul>
        </div>
      </nav>
    </footer>
  </main>
</DocumentFragment>
`;
